solutions/usaco-994.mdx

```diff
@@ -2,49 +2,134 @@
 id: usaco-994
 source: USACO Gold 2020 January
 title: Farmer John Solves 3SUM
-author: Qi Wang
+author: Qi Wang, Ryan Chou
 ---
 
 [Official Analysis (C++ and Java)](http://www.usaco.org/current/data/sol_threesum_gold_jan20.html)
 
+## Explanation
+
+Although the problem may be daunting at first, it helps to think about how a smaller range is connected to another.
+
+If $\texttt{ways}[i][j]$ equals the number of unordered triplets which sum to zero between $[i, j]$, then what states does this rely on?
+
+![](usaco-994/init.png)
+
+Well intuitively, we can choose to exclude one endpoint or the other, which gives us a picture like this (reminder to make sure not to double count the overlap).
+
+![](usaco-994/ps.png)
+
+While this'll count all the triplets in between them, we aren't taking into account any triplets which use both the elements at $i$ and $j$, since those aren't included in our "PIE-like" transition.
+
+To do this, we can precalculate the number of triplets which start at $i$ and end at $j$ in $\mathcal{O}(N^2)$ time by storing the occurences of the compliment of $i$ and $j$ in an array.
+
+More formally, our final transition is:
+
+$\texttt{ways}[i][j] = \texttt{ways}[i + 1][j] + \texttt{ways}[i][j - 1] - \texttt{ways}[i + 1][j - 1] + \texttt{trp}[i][j]$
+
+where $\texttt{trp}[i][k]$ equals the number of triplets which sum to zero and start at $i$ and end at $k$.
+
+Note that we can't store $\texttt{trp}$ in its own array due to memory constraints.
+
+
 ## Implementation
 
 **Time Complexity:** $\mathcal{O}(N^2)$
 
 <LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+const int MAX_VAL = 1e6;
+
+int main() {
+	freopen("threesum.in", "r", stdin);
+	freopen("threesum.out", "w", stdout);
+	cin.tie(0)->sync_with_stdio(0);
```

Comment:
```suggestion
	cin.tie(0) -> sync_with_stdio(0);
```
looks more fitting

Reply:
don't think that's how it works
the `->` operator doesn't have spaces, it's almost like `.`

Reply:
spaces doesnt matter

Reply:
What

Reply:
what

Reply:
not rly sure it matters? ill commit this time though